cmd/tools/gensearchattributehelpers/main.go

```diff
@@ -0,0 +1,183 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path"
+	"reflect"
+	"regexp"
+	"strings"
+	"text/template"
+
+	historypb "go.temporal.io/api/history/v1"
+)
+
+type (
+	eventData struct {
+		EventTypeCaps    string
+		AttributesGetter string
+	}
+
+	searchAttributesHelpersData struct {
+		Events []eventData
+	}
+)
+
+var (
+	// Is used to find attribute getters and extract the event type (match[1]).
+	attributesGetterRegex = regexp.MustCompile("^Get(.+)EventAttributes$")
+)
+
+func main() {
+	outPathFlag := flag.String("out", ".", "path to write generated files")
+	licenseFlag := flag.String("copyright_file", "LICENSE", "path to license to copy into header")
+	flag.Parse()
+
+	licenseText := readLicenseFile(*licenseFlag)
+
+	eventHelpersFile := path.Join(*outPathFlag, "event_gen.go")
+	callWithFile(generateSearchAttributesEventHelpers, eventHelpersFile, licenseText)
+}
+
+func generateSearchAttributesEventHelpers(w io.Writer) {
+
+	writeSearchAttributesEventHelpers(w, `
+package searchattribute
+
+import (
+	commonpb "go.temporal.io/api/common/v1"
+	enumspb "go.temporal.io/api/enums/v1"
+	historypb "go.temporal.io/api/history/v1"
+)
+
+func SetToEvent(event *historypb.HistoryEvent, sas *commonpb.SearchAttributes) bool {
+	switch event.GetEventType() {
+	{{- range .Events}}
+	case enumspb.EVENT_TYPE_{{.EventTypeCaps}}:
+		event.{{.AttributesGetter}}().SearchAttributes = sas
+		return true
+	{{- end}}
+	default:
+		return false
+	}
+}
+
+func GetFromEvent(event *historypb.HistoryEvent) (*commonpb.SearchAttributes, bool) {
+	switch event.GetEventType() {
+	{{- range .Events}}
+	case enumspb.EVENT_TYPE_{{.EventTypeCaps}}:
+		return event.{{.AttributesGetter}}().GetSearchAttributes(), true
+	{{- end}}
+	default:
+		return nil, false
+	}
+}
+`)
+}
+
+func writeSearchAttributesEventHelpers(w io.Writer, tmpl string) {
+	sahd := searchAttributesHelpersData{}
+
+	historyEventT := reflect.TypeOf((*historypb.HistoryEvent)(nil))
+
+	for i := 0; i < historyEventT.NumMethod(); i++ {
+		attributesGetter := historyEventT.Method(i)
+		matches := attributesGetterRegex.FindStringSubmatch(attributesGetter.Name)
+		if len(matches) < 2 {
+			continue
+		}
+		if attributesGetter.Type.NumOut() != 1 {
+			continue
+		}
+		if _, found := attributesGetter.Type.Out(0).MethodByName("GetSearchAttributes"); !found {
+			continue
+		}
+
+		ed := eventData{
+			EventTypeCaps:    camelCaseToScreamingSnakeCase(matches[1]),
+			AttributesGetter: attributesGetter.Name,
+		}
+		sahd.Events = append(sahd.Events, ed)
+	}
+
+	fatalIfErr(template.Must(template.New("code").Parse(tmpl)).Execute(w, sahd))
+}
+
+func callWithFile(generator func(io.Writer), outFile string, licenseText string) {
```

Comment:
@dnr, I swear, I will extract it after this PR. I didn't actually expect to use code generation here.

Reply:
When all you have a hammer...